File: 2017/day3.py
puzzle_input = int(input("Puzzle input? "))


# Part 1
def ring_size(ring_num):
    if ring_num == 0:
        return 1
    return 8 * ring_num


def ring_max(ring_num):
    return (2 * ring_num + 1) ** 2 - 1


def ring_min(ring_num):
    return ring_max(ring_num) - ring_size(ring_num) + 1


def edge_distance(ring_num):
    return ring_size(ring_num) // 4


def ring_cardinal_index(ring_num, direction):
    if not (0 <= direction <= 3):
        raise Exception("Bad direction")
    return ring_min(ring_num) + ring_num - 1 + direction * edge_distance(ring_num)


def get_distance(index):
    # memory index is 1 based
    index -= 1

    ring_num = 0
    while ring_max(ring_num) < puzzle_input:
        ring_num += 1

    cardinals = (ring_cardinal_index(ring_num, dir) for dir in range(4))
    differences = (abs(cardinal - index) for cardinal in cardinals)
    ring_distance = min(differences)

    total_distance = ring_num + ring_distance
    return total_distance


print("input {} is distance {}".format(puzzle_input, get_distance(puzzle_input)))

# Part 2
import math

from typing import Iterable


class SpiralMemory(object):
    def __init__(self, memory: Iterable[int]):
        self.memory = list(memory)

    def __getitem__(self, key):
        x, y = key
        index = self.get_index(x, y)
        if index >= len(self.memory):
            return 0
        else:
            return self.memory[index]

    def __setitem__(self, key, value):
        row_off, col_off = key
        index = self.get_index(x, y)
        if index >= len(self.memory):
            self.memory.extend([0] * (index + 1 - len(self.memory)))
        self.memory[index] = value

    def get_index(self, x, y):

        if y == 0 and x == 0:
            return 0

        ring = max(abs(y), abs(x))
        if abs(y) < abs(x):
            ring_index = y
        else:
            ring_index = x

        start_index = 1
        for i in range(1, ring):
            start_index += ring_size(i)

        # if they are equal, let the y>=x case handle it to deal with the bottom right corner correctly
        if abs(x) > abs(y):
            if x > 0:
                east_index = start_index + ring - 1
                assert (east_index + ring_index < start_index + ring_size(ring))
                return east_index + ring_index
            else:
                west_index = start_index + ring - 1 + 2 * edge_distance(ring)
                assert (west_index + ring_index < start_index + ring_size(ring))
                return west_index - ring_index
        elif abs(y) >= abs(x):
            if y > 0:
                north_index = start_index + ring - 1 + edge_distance(ring)
                assert (north_index + ring_index < start_index + ring_size(ring))
                return north_index - ring_index
            else:
                south_index = start_index + ring - 1 + 3 * edge_distance(ring)
                assert (south_index + ring_index < start_index + ring_size(ring))
                return south_index + ring_index

    def get_offsets(self, index):
        if index == 0:
            return (0, 0)

        ring_start = 1
        ring = 1
        while index >= ring_start + ring_size(ring):
            ring_start += ring_size(ring)
            ring += 1

        ring_index = index - ring_start
        east_index = ring - 1
        north_index = east_index + edge_distance(ring)
        west_index = north_index + edge_distance(ring)
        south_index = west_index + edge_distance(ring)
        dist = lambda cardinal_index: abs(ring_index - cardinal_index)
        min_cardinal_dist = min(dist(east_index), dist(north_index), dist(west_index), dist(south_index))

        if abs(ring_index - east_index) <= edge_distance(ring) // 2:
            return (ring, ring_index - east_index)
        elif abs(west_index - ring_index) <= edge_distance(ring) // 2:
            return (-ring, west_index - ring_index)
        elif abs(north_index - ring_index) <= edge_distance(ring) // 2:
            return (north_index - ring_index, ring)
        elif abs(ring_index - south_index) <= edge_distance(ring) // 2:
            return (ring_index - south_index, -ring)

    def print_me(self, ring=None):
        if ring is None:
            ring = max(abs(o) for o in self.get_offsets(len(self.memory) - 1))

        mem_size = max(self.get_index(ring, -ring) + 1, len(self.memory))
        digits = math.ceil(math.log10(max(self.memory[:mem_size])))
        format_str = "{{:{}d}}".format(digits)
        for y in range(ring, -ring - 1, -1):
            print(" ".join((format_str.format(mem[x, y]) for x in range(-ring, ring + 1))))


mem = SpiralMemory([1])

value = 0
i = 0
while value < puzzle_input:
    i += 1
    x, y = mem.get_offsets(i)
    # now sum in adjacent
    sum = 0
    for x_off in (-1, 0, 1):
        for y_off in (-1, 0, 1):
            # value is already zero, so adding in 0,0 is benign
            # And now that I've explained that, I could have already done it in the length of these two comments..
            sum += mem[x + x_off, y + y_off]

    mem[x, y] = sum
    value = sum

mem.print_me()
print("first value greater than {} is {}".format(puzzle_input, value))

File: rust/2017/src/bin/day3.rs
fn ring_size(ring_num: i32) -> i32 {
    assert!(ring_num >= 0, "Ring number must be positive");
    match ring_num {
        0 => 1,
        _ => 8 * ring_num,
    }
}

fn ring_max(ring_num: i32) -> i32 {
    (2 * ring_num + 1).pow(2)
}

fn ring_min(ring_num: i32) -> i32 {
    ring_max(ring_num - 1) + 1
}

fn edge_distance(ring_num: i32) -> i32 {
    ring_size(ring_num) / 4
}

#[derive(Copy, Clone)]
enum Direction {
    Right = 0,
    Up = 1,
    Left = 2,
    Down = 3,
}

fn ring_cardinal_index(ring_num: i32, direction: Direction) -> i32 {
    let right_cardinal = ring_min(ring_num) + ring_num - 1;
    right_cardinal + direction as i32 * edge_distance(ring_num)
}

// Part 1
fn distance_from_center(square_num: i32) -> i32 {
    // memory index is 1-based
    let ring_num = (((square_num as f32).sqrt() - 1.0) / 2.0).ceil() as i32;

    use Direction::*;
    let ring_distance = [Up, Down, Left, Right].iter()
                                               .map(|dir| ring_cardinal_index(ring_num, *dir))
                                               .map(|cardinal_index| (cardinal_index - square_num).abs())
                                               .min().unwrap();

    ring_num + ring_distance
}

// Part 2
// Todo: Add predicate
fn stress_test(_value: i32) -> i32 {
    unimplemented!();
}

#[test]
fn test_part1() {
    assert_eq!(distance_from_center(1), 0);
    assert_eq!(distance_from_center(12), 3);
    assert_eq!(distance_from_center(23), 2);
    assert_eq!(distance_from_center(1024), 31);
}

/*
#[test]
fn test_part2() {
    assert_eq!(stress_test_value(1), 1);
    assert_eq!(stress_test_value(2), 1);
    assert_eq!(stress_test_value(3), 2);
    assert_eq!(stress_test_value(4), 4);
    assert_eq!(stress_test_value(5), 5);
    assert_eq!(stress_test_value(6), 10);
    assert_eq!(stress_test_value(7), 11);
    assert_eq!(stress_test_value(8), 23);
    assert_eq!(stress_test_value(9), 25);
    assert_eq!(stress_test_value(10), 26);
    assert_eq!(stress_test_value(11), 54);
    assert_eq!(stress_test_value(12), 57);
    assert_eq!(stress_test_value(13), 59);
    assert_eq!(stress_test_value(14), 122);
    assert_eq!(stress_test_value(15), 133);
    assert_eq!(stress_test_value(16), 142);
    assert_eq!(stress_test_value(17), 147);
    assert_eq!(stress_test_value(18), 304);
    assert_eq!(stress_test_value(19), 330);
    assert_eq!(stress_test_value(20), 351);
    assert_eq!(stress_test_value(21), 362);
    assert_eq!(stress_test_value(22), 747);
    assert_eq!(stress_test_value(23), 806);
}
*/

fn main() {
    let input = 361527;

    let part1_answer = distance_from_center(input);
    println!("Part 1: {}", part1_answer);

    let part2_answer = stress_test(input);
    println!("Part 2: {}", part2_answer);
}